sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -267,6 +262,10 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
       provider.getTable(dsOptions) match {
         case table: SupportsWrite if table.supports(BATCH_WRITE) =>
+          if (partitioningColumns.nonEmpty) {
```

Comment:
good catch! Even if the format is a `TableProvider`, we may still fall back to v1. It's better to check the partition when we are really going to do a v2 write.

Reply:
BTW, technically we only need to assert no partition columns for append and overwrite. Since the v2 write here only supports append and overwrite, we can revisit it later.